### ISSUE TITLE
ImpureMethodCall false positive

### ISSUE THREAD

**Issue by iluuu1994** [vimeo/psalm#2090]
```php
abstract class Foo {
  private array $options;
  private array $defaultOptions;
  
  function __construct(array $options) {
    $this->setOptions($options);
    $this->setDefaultOptions($this->getOptions());
  }

  function getOptions(): array {
    return $this->options;
  }
  
  function setOptions(array $options): void {
    $this->options = $options;
  }

  function setDefaultOptions(array $defaultOptions): void {
    $this->defaultOptions = $defaultOptions;
  }
}
```

> ERROR: ImpureMethodCall - 9:12 - Cannot call an possibly-mutating method Foo::setdefaultoptions from a mutation-free context

https://psalm.dev/r/4bf732d5eb

Psalm gets confused here. Changing line 9 to `$this->setDefaultOptions($options);` works just fine: https://psalm.dev/r/cc101a49d7

**EDIT**

Well, it says "possibly-mutating". Is this expected?

**Comment by muglug** [comment]
Fixed in f096c3d — I forgot to ensure that the purity checks don’t kick in when checking for property initialisation.

**Comment by iluuu1994** [comment]
Thank you @muglug!

**Comment by iluuu1994** [comment]
I tried it again with f20dc16295f15eb79a752f135e02f8dd6d0e38d3 and most of the errors are indeed gone. I still get 3 `ImpurePropertyAssignment` false positives. No lock with reproducing them yet.

**Comment by muglug** [comment]
If you predicate `ImpurePropertyAssignment` issues being emitted on

```
if (!$context->collect_mutations && !$context->collect_initializations) {
```

do the false positives disappear?

**Comment by iluuu1994** [comment]
Changing this line:

https://github.com/vimeo/psalm/blob/f20dc16295f15eb79a752f135e02f8dd6d0e38d3/src/Psalm/Internal/Analyzer/Statements/Expression/AssignmentAnalyzer.php#L571

to:

```php
if ($context->mutation_free && !$context->collect_mutations && !$context->collect_initializations) {
```

indeed fixes the issue.

**Comment by muglug** [comment]
Patched in cfdfb9b

**Comment by iluuu1994** [comment]
It's working now, thank you!
